Upgrade to MUI v5

cmdcolin opened the pull request:
This is a draft PR to upgrade to using MUI v5. A large change, but generally is approaching working state


There is a small amount of back-compatibility for plugins that might use MUIv4 by using ReExports to export the @mui/material namespace into @material-ui/core, so plugins using v4 with @material-ui/core can still load (MSAView is an example of this and it still loads)

Some benefits
- Newer ecosystem won't go stale
- This PR would help us avoid pinning to webpack v5.70.0 (required by current usage of data-grid in UMD build of embedded components, see the "returntrue" syntax error https://github.com/mui/mui-x/issues/4490), sort of related to ecosystem going stale

Some remaining todos

- storybook not loading currently AFAIK
- typescripting errors related to mui/styles conversion

Example link

https://jbrowse.org/code/jb2/mui5/?config=test_data%2Fvolvox%2Fconfig.json&session=share-V0DO6dUvkA&password=qsULt

carolinebridge-oicr commented:
![image](https://user-images.githubusercontent.com/83305007/167908372-2631382c-e94c-4da2-b0b2-5cda633b4656.png)

just briefly clicking around, mini controls are comically large

cmdcolin commented:
this is nearly ready, i can mark it ready for review

some notable changes

1) removes @material-ui/lab from reexports. these are unused by the main codebase now and change a lot anyways
2) uses tss-react for styling, which allows us to retain the makeStyles usage https://github.com/garronej/tss-react (their tagline: makeStyles is dead, long live makeStyles!). i used an automated the conversion with a codemod https://mui.com/material-ui/migration/migrating-from-jss/ so it is semi-blessed by the MUI team
3) using `yarn start` in jbrowse-web gives some various warnings about declare.d.ts things that seem like false positives that are not on main, but appear harmless

overall this touches lots of code files, but i mostly made it so that the UI on this branch is the same as what is on main




links
https://jbrowse.org/storybook/lgv/mui5/?path=/story/getting-started--page
https://jbrowse.org/storybook/cgv/mui5/?path=/story/getting-started--page

web
https://jbrowse.org/code/jb2/mui5/?config=test_data%2Fconfig_demo.json&session=share-ldgNeRs2Ie&password=6QTSO

also note: merging this could involve a major version bump if we want to be safe. we make a best effort at back compatibility with ReExports but it is hard to say if this is truly backwards compatible

cmdcolin commented:
also just to contextualize, I ran into the `returntrue` issue making the nextjs demo #3022 which is related to @mui/x-data-grid on mui v4, and i think probably the best way to fix that is just upgrading dependencies to v5. that is a somewhat specific thing with a workaround (stay on older version of nextjs) but hopefully this PR helps to reduce that sort of lock in

cmdcolin commented:
would be interested in thoughts on this PR plus https://github.com/GMOD/jbrowse-components/pull/3024 and https://github.com/GMOD/jbrowse-components/pull/3029 (kinda a trifecta :))

cmdcolin commented:
@garrettjstevens if you want to try all the upgrades (this branch + react 18 + lazy reexports) can try https://github.com/GMOD/jbrowse-components/tree/mui5_allupgrades

garrettjstevens commented:
I tried running the Apollo plugin under this branch. It wouldn't load at first, and I had to make this change to get it to load:
```diff
diff --git a/packages/core/ui/theme.ts b/packages/core/ui/theme.ts
--- a/packages/core/ui/theme.ts
+++ b/packages/core/ui/theme.ts
@@ -190,14 +190,19 @@ export function createJBrowseTheme(theme?: any) {
   if (theme?.palette?.tertiary) {
     theme = deepmerge(theme, {
       palette: {
-        tertiary: refTheme.palette.augmentColor(theme.palette.tertiary),
+        tertiary: refTheme.palette.augmentColor({
+          color: theme.palette.tertiary,
+        }),
       },
     })
   }
   if (theme?.palette?.quaternary) {
+    console.log(theme.palette.quaternary)
     theme = deepmerge(theme, {
       palette: {
-        quaternary: refTheme.palette.augmentColor(theme.palette.quaternary),
+        quaternary: refTheme.palette.augmentColor({
+          color: theme.palette.quaternary,
+        }),
       },
     })
   }
```
It loaded then, but the first screen looks like this:

![localhost_3000__config=http%3A%2F%2Flocalhost%3A9000%2Fjbrowse_config json](https://user-images.githubusercontent.com/25592344/174339800-32ce3994-0e13-4731-be02-24de709e8678.png)

On main it looks like this:

![localhost_3000__config=http%3A%2F%2Flocalhost%3A9000%2Fjbrowse_config json (1)](https://user-images.githubusercontent.com/25592344/174340414-70311d5b-53a8-40ff-8c4c-dac808bd1883.png)

If I click on "Collaborative", it then crashes with this error:

```
Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `AccountCard`.
```

I could look more into that error and see what component is causing it, but that'll probably be next week sometime.

cmdcolin commented:
nice, I added a couple I think things that were missing re-exports for List and Avatar (all re-exports have to be named imports so I just went through node_modules/@mui/material/index.js incrementally to get them) and added that theme patch. I am not sure about the uncentered stuff. I could also check it out too if I can get the plugin setup

garrettjstevens commented:
Looks like that error was fixed, but now I'm getting this one:
```
Button.js:104 Uncaught TypeError: Cannot read properties of undefined (reading 'dark')
    at ButtonRoot.ownerState.ownerState (Button.js:104:1)
    at transformedStyleArg (createStyled.js:191:1)
    at handleInterpolation (emotion-serialize.browser.esm.js:139:1)
    at serializeStyles (emotion-serialize.browser.esm.js:264:1)
    at emotion-styled-base.browser.esm.js:131:1
    at emotion-element-cbed451f.browser.esm.js:36:1
    at renderWithHooks (react-dom.development.js:16305:1)
    at updateForwardRef (react-dom.development.js:19226:1)
    at beginWork (react-dom.development.js:21636:1)
    at HTMLUnknownElement.callCallback (react-dom.development.js:4164:1)
```
I'm not sure what that's from. If you want to check it out, that step doesn't need an Apollo server set up so you can start up the plugin by:

1. Starting up jbrowse-web on port 3000
2. Clone the [Apollo3](https://github.com/GMOD/Apollo3) repo and do:
```
yarn
cd packages/jbrowse-plugin-apollo/
yarn start
```
3. Go to http://localhost:3000/?config=http%3A%2F%2Flocalhost%3A9000%2Fjbrowse_config.json

garrettjstevens commented on packages/core/ReExports/modules.tsx:
```diff
@@ -0,0 +1,280 @@
+  Dialog: lazy(() => import('@mui/material/Dialog')),
+  DialogActions: lazy(() => import('@mui/material/DialogActions')),
+  DialogTitle: lazy(() => import('@mui/material/DialogTitle')),
+  DialogContent: lazy(() => import('@mui/material/DialogContent')),
```
```suggestion
  DialogContent: lazy(() => import('@mui/material/DialogContent')),
  DialogContentText: lazy(() => import('@mui/material/DialogContentText')),
```

cmdcolin commented:
awesome :partying_face: really glad to hear it works with apollo :)
